modules/compiler/utils/source/mux_builtin_info.cpp

```diff
@@ -1127,5 +1127,49 @@ Value *BIMuxInfoConcept::initializeSchedulingParamForWrappedKernel(
                         /*ArraySize*/ nullptr, Info.ParamName);
 }
 
+std::optional<llvm::ConstantRange> BIMuxInfoConcept::getBuiltinRange(
+    llvm::CallInst &CI, BuiltinID ID,
+    std::array<std::optional<uint64_t>, 3> MaxLocalSizes,
+    std::array<std::optional<uint64_t>, 3> MaxGlobalSizes) const {
+  assert(CI.getCalledFunction() && CI.getType()->isIntegerTy() &&
+         "Unexpected builtin");
+
+  auto Bits = CI.getType()->getIntegerBitWidth();
+  // Assume we're indexing the global sizes array.
+  std::array<std::optional<uint64_t>, 3> *SizesPtr = &MaxGlobalSizes;
+
+  switch (ID) {
+    default:
+      return std::nullopt;
+    case eMuxBuiltinGetWorkDim:
+      return ConstantRange::getNonEmpty(APInt(Bits, 1), APInt(Bits, 4));
+    case eMuxBuiltinGetLocalId:
+    case eMuxBuiltinGetLocalSize:
+    case eMuxBuiltinGetEnqueuedLocalSize:
+      // Use the local sizes array, and fall through to common handling.
+      SizesPtr = &MaxLocalSizes;
+      [[fallthrough]];
+    case eMuxBuiltinGetGlobalSize: {
+      auto *DimIdx = CI.getOperand(0);
+      if (!isa<ConstantInt>(DimIdx)) {
+        return std::nullopt;
+      }
+      uint64_t DimVal = cast<ConstantInt>(DimIdx)->getZExtValue();
+      if (DimVal >= SizesPtr->size() || !(*SizesPtr)[DimVal]) {
+        return std::nullopt;
+      }
+      // ID builtins range from [0,size) and size builtins from [1,size]. Thus
```

Comment:
I know these are mathematical descriptions, but not everyone remembers their meaning so I think I'd prefer it to be described e.g. 0 to size-1 and 1 to size inclusive

Reply:
I think I'll keep the concise notation but add `(inclusive)` and `(exclusive)` after each to be extra clear.